### PR TITLE
secboot: use half the mem for KDF in AddRecoveryKey

### PR DESCRIPTION
**Pull request by mvo5**
Instead of benchmarking the KDF parameters for the recovery key
(which takes some time to run) we can also use defaults for the
KDF parameters. The defaults suggested by Chris are "4 iterations"
and half the usable memory. This commit implements the suggestions.

### CODE REVIEW

**Review by anonymouse64 — APPROVED** [review]
lgtm, one suggestion about how to avoid the pointer usage just because we can avoid it and I think avoiding it makes the code more readable

File: osutil/meminfo.go
Comment (mardy) [reply osutil/meminfo.go:37]: and usabable -> usable :-)

File: osutil/meminfo.go
Comment (mardy) [inline osutil/meminfo.go:77]: Does it make sense to add a check to ensure that `memTotal >= cmaTotal`, since the type of the value returned by this function is unsigned?

I doubt that memInfo would return such bogus values, but...

**Comment by mvo5** [comment]
This needs to be re-run with "run-nested" before it its merged. But let's wait for feedback from @chrisccoulson first.

File: secboot/encrypt_sb.go
Comment (anonymouse64) [inline secboot/encrypt_sb.go:71]:
```suggestion
	// usable memory and subtracting hardcoded 384MB that is
```

File: secboot/encrypt_sb.go
Comment (anonymouse64) [inline secboot/encrypt_sb.go:87]: all of this would be more readable IMHO if we used the gadget.Size type here, but not necessary right now given the time pressure to try this out

File: osutil/meminfo.go
Comment (anonymouse64) [inline osutil/meminfo.go:41]: also again ideally this would return a quantity.Size instead of a uint64

File: secboot/encrypt_sb.go
Comment (anonymouse64) [inline secboot/encrypt_sb.go:81]:
```suggestion
	// min 32 KB
	if kdfMem < 32*1024 {
```

**Review by anonymouse64 — COMMENTED** [review]
some nitpicks, mainly about the fact that there are an awful lot of manual numbers which can sometimes be confusing, if possible I would suggest using unit / types instead so it's more obvious what unit each number is for etc.

**Comment by mvo5** [comment]
I got confirmation that this approach helps and the whole boot time decreased from ~187s to ~165s during the install mode on a reference board we use. So this seems to be good to go.

File: secboot/encrypt_sb.go
Comment (mvo5) [reply secboot/encrypt_sb.go:65]: I added a small comment, please let me know if this looks sufficient

**Comment by mvo5** [comment]
Thanks for the excellent comments. I'm inclined to start with this minimal version for 2.51 and then do a followup that moves to using the "quantity" pacakge. WDYT? Especially @pedronis and @chrisccoulson ? Your review would be really appreciated before we get this out.

File: secboot/encrypt_sb.go
Comment (anonymouse64) [reply secboot/encrypt_sb.go:65]: a bit hand wavy, but good enough :smile:

**Comment by anonymouse64** [comment]
> I'm inclined to start with this minimal version for 2.51 and then do a followup that moves to using the "quantity" pacakge

fine with me

File: secboot/encrypt_sb.go
Comment (pedronis) [inline secboot/encrypt_sb.go:76]: do we know how much memory we actually have free on the device during snap-bootstrap and during install mode? 384 is kind of for the full user space run mode? just wondering if it's a bit if it's too pessimistic for here.

File: secboot/encrypt_sb_test.go
Comment (pedronis) [inline secboot/encrypt_sb_test.go:60]: didn't we discuss that this also should become just 32?

**Review by pedronis — COMMENTED** [review]
two questions, one about the 384 factor we can consider tweaking later on master.

File: secboot/encrypt_sb_test.go
Comment (mvo5) [reply secboot/encrypt_sb_test.go:60]: Indeed, I was wondering and it seems it's best if I open a separate PR for this PR is about the recovery key and this change is about the main key. So unless you ffeel differently I do a small extra PR with the 32mb->32kb change for the main key.

Comment (mvo5) [reply secboot/encrypt_sb_test.go:60]: I did this now in https://github.com/snapcore/snapd/pull/10645

Comment (pedronis) [reply secboot/encrypt_sb_test.go:60]: that's fine, thanks

File: secboot/encrypt_sb.go
Comment (pedronis) [reply secboot/encrypt_sb.go:76]: as I said, I don't think it's a blocker but we might want to explore a bit more what to do on master about this

**Review by pedronis — APPROVED** [review]
+1 taking into account the answers to the questions

File: osutil/meminfo.go
Comment (pedronis) [reply osutil/meminfo.go:77]: I wonder what happens with this computation on devices with something like android ION (see the PR making an interface for that), I suppose we will have to find out

**Pull request merged**
